micro-rdk-nmea/src/parse_helpers/parsers.rs

```diff
@@ -0,0 +1,404 @@
+use bitvec::prelude::*;
+use std::{
+    marker::PhantomData,
+    rc::Rc,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+use micro_rdk::common::sensor::GenericReadingsResult;
+
+use super::{
+    enums::NmeaEnumeratedField,
+    errors::{NmeaParseError, NumberFieldError},
+};
+
+/// Cursor over a byte slice of data. The cursor can be moved by either consuming
+/// child-instances of `DataRead` or calling the `advance` function
+pub struct DataCursor<'a> {
+    data: &'a [u8],
+    bit_position: Rc<AtomicUsize>,
+}
+
+struct DataRead<'a> {
+    data: &'a [u8],
+    bit_position: Rc<AtomicUsize>,
+    bit_size: usize,
+}
+
+impl<'a> Drop for DataRead<'a> {
+    fn drop(&mut self) {
+        self.bit_position.fetch_add(self.bit_size, Ordering::SeqCst);
+    }
+}
+
+impl<'a> DataCursor<'a> {
+    pub fn new(data: &'a [u8]) -> Self {
+        Self {
+            data,
+            bit_position: Rc::new(AtomicUsize::new(0)),
+        }
+    }
+
+    fn read(&self, bits: usize) -> Result<DataRead, NmeaParseError> {
+        let bit_position = self.bit_position.load(Ordering::SeqCst);
+        let end_byte_position = (bit_position + bits).div_ceil(8);
+        if end_byte_position > self.data.len() {
+            Err(NmeaParseError::EndOfBufferExceeded)
+        } else {
+            let start_byte_position = bit_position / 8;
+            Ok(DataRead {
+                data: &self.data[start_byte_position..end_byte_position],
+                bit_position: self.bit_position.clone(),
+                bit_size: bits,
+            })
+        }
+    }
+
+    pub fn advance(&self, bits: usize) -> Result<(), NmeaParseError> {
```

Comment:
Could `advance` be implemented by calling `read` and throwing away the result?

Reply:
It could, is that preferred?

Reply:
I think so, because then `advance` is very very simple, and the complexity lives only in `read`.